File: src/contests/cg/puzzle/rust/euclids-algorithm.rs
// 🎮 CodinGame Puzzle - euclids-algorithm
// https://www.codingame.com/training/easy/euclids-algorithm

fn rl() -> String {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

fn main() {
    let parts: Vec<i64> = rl().split_whitespace().map(|x| x.parse().unwrap()).collect();
    let (a, b) = (parts[0], parts[1]);
    let (mut x, mut y) = (a, b);
    while y != 0 {
        let q = x / y;
        let r = x % y;
        println!("{}={}*{}+{}", x, y, q, r);
        x = y;
        y = r;
    }
    println!("GCD({},{})={}", a, b, x);
}

File: src/contests/cg/puzzle/ocaml/euclids-algorithm.ml
(* 🎮 CodinGame Puzzle - euclids-algorithm *)
(* https://www.codingame.com/training/easy/euclids-algorithm *)

let () =
  let parts = String.split_on_char ' ' (String.trim (read_line ())) in
  let a = int_of_string (List.nth parts 0) in
  let b = int_of_string (List.nth parts 1) in
  let x = ref a and y = ref b in
  while !y <> 0 do
    let q = !x / !y and r = !x mod !y in
    Printf.printf "%d=%d*%d+%d\n" !x !y q r;
    x := !y;
    y := r
  done;
  Printf.printf "GCD(%d,%d)=%d\n" a b !x
